### PR TITLE
Domain zone controller

### PR DESCRIPTION
**Pull request by DeRain**

### CODE REVIEW

File: contracts/controllers/DomainZoneController.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        _registry = registry;
+        for (uint256 index = 0; index < accounts.length; index++) {
+            _addWhitelisted(accounts[index]);
+        }
+    }
+
+    function mintChild(address to, uint256 tokenId, string memory label, string[] memory keys, string[] memory values) public onlyWhitelisted {
+        address resolver = _registry.resolverOf(tokenId);
+        uint256 childTokenId = _registry.childIdOf(tokenId, label);
+        _registry.mintChild(address(this), tokenId, label);
```
Comment (bogdan) [inline contracts/controllers/DomainZoneController.sol:28]: Add optimization if there is no preconfiguration (keys are empty), just `mintChild` to the final owner without reassign and preconfiguration.

File: contracts/controllers/DomainZoneController.sol [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        _registry.resolveTo(resolver, childTokenId);
+        IResolver(resolver).reconfigure(keys, values, childTokenId);
+        _registry.setOwner(to, childTokenId);
+        emit MintChild(childTokenId);
```
Comment (bogdan) [inline contracts/controllers/DomainZoneController.sol:32]: We need more info here. I would also include `parentTokenId` and `label` to make domain name recoverable.

File: test/BulkWhitelistedRole.ganache.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    await bulkWhitelistedRole.bulkRemoveWhitelisted(accounts);
+    for (account of accounts) {
+        assert.isFalse(await bulkWhitelistedRole.isWhitelisted(account))
+    }
+  })
+
+  it('should not allow add or remove accounts from non-admin address', async () => {
+    try {
+        await bulkWhitelistedRole.bulkAddWhitelisted(accounts, {from: notAdmin});
+        assert.fail('bulkAddWhitelisted function should fail when trying to call from not allowed address')
```
Comment (mvlabat) [inline test/BulkWhitelistedRole.ganache.js:32]: I don't know how chai `assert.fail` works, but I didn't find it in the docs that it's not recoverable. Maybe `assert.throws` would be a cleaner way to check this?

File: test/DomainZoneController.ganache.js [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+  it('addresses added in constructor should be whitelisted', async () => {
+    const domainZoneController = await DomainZoneController.new(registry.address, accounts)
+    assert.isAbove(accounts.length, 0)
+    for (account of accounts) {
+        assert.isTrue(await domainZoneController.isWhitelisted(account))
+    }
+  })
+
+  it('address should not be whitelisted if weren\'t added', async () => {
```
Comment (mvlabat) [inline test/DomainZoneController.ganache.js:45]: nitpick: wasn't

File: test/DomainZoneController.ganache.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    await registry.approve(domainZoneController.address, secondLevelTokenId)
+    await domainZoneController.mintChild(domainReceiver, secondLevelTokenId, subdomainName, [], [], {from: whitelisted})
+    const subdomainTokenId = await registry.childIdOf(secondLevelTokenId, subdomainName)
+    assert.equal(
+        await registry.ownerOf(subdomainTokenId),
+        domainReceiver
+    )
+  })
+
+  it('should not allow minting from not allowed second-level domains', async () => {
```
Comment (mvlabat) [inline test/DomainZoneController.ganache.js:103]: Not a comment, but a question (for my education): what does "not allowed second-level domain" mean? What condition makes them being allowed?

File: test/Resover.geth.js [hunk 6]
```diff
@@ -20,14 +21,18 @@ contract('Resolver', function([coinbase, notOwner]) {
...
+
   beforeEach(async () => {
     registry = await Registry.deployed()
     mintingController = await MintingController.deployed()
     resolver = await Resolver.deployed()
   })
 
   it('should reconfigure resolver with new values', async () => {
-    const tok = await initializeDomain('reconfigure')
+    const tok = (await initializeDomain('reconfigure')).toString(10)
```
Comment (mvlabat) [inline test/Resover.geth.js:35]: what's tok?

File: test/BulkWhitelistedRole.ganache.js (hunk 3 above)
Comment (DeRain) [reply test/BulkWhitelistedRole.ganache.js:32]: @mvlabat I need to check the custom field in error object. Unfortunately, I didn't found how to do this via `assert.throws`

File: test/DomainZoneController.ganache.js (hunk 5 above)
Comment (DeRain) [reply test/DomainZoneController.ganache.js:103]: @mvlabat you will need to transfer or approve your second-level domain before `DomainZoneController` will be able to use it and mint subdomain child.

File: test/Resover.geth.js (hunk 6 above)
Comment (DeRain) [reply test/Resover.geth.js:35]: @mvlabat BigNumber instance.

**Pull request merged**